### DISCUSSION TITLE
V0.6 cost more memory than V0.5.1 with frequent update on DB ?

### DISCUSSION THREAD

**Discussion by foTok** [duckdb/duckdb#5626] [Q&A]
I made some analysis on a frequently updated DB, and found that memory increasing of v0.6 was much faster than v0.5.1.
Although DuckDB's memory_usage was 1.6 GB, the programs cost  more than 8.6 GB.
Why so much extra memory was used ?

DuckDB info:
```
2022-12-07 23:56:53
database_size	block_size	total_blocks	used_blocks	free_blocks	wal_size	memory_usage	memory_limit	
VARCHAR	BIGINT	BIGINT	BIGINT	BIGINT	VARCHAR	VARCHAR	VARCHAR	
[ Rows: 1]
543.4MB	262144	2073	1764	309	250.6MB	1.6GB	26.8GB
```

Shell info:
```
2022-12-07 23:56:39
DuckDBVerification VSZmem: 8636344 VmRSS: 7847320 kB CPU: 312
```

**Comment by Mytherin** [comment]
It is impossible to tell without knowing anything about what you are doing.

**Reply by foTok** [reply]
Sorry for the inexplicit. My case is as follows:
1. Create a DB in file mode, create some table and load some original data.
2. Start a thread, which does data analysis in a while(true) loop: the source tables are stored in files, and the results are stored in a temporary table, the temporary table is destroy by the connection at the end of each loop.
3. Start a thread, which randomly insert and delete data in tables stored in file (also in a while loop): I always keep the insert number equals to the delete number, so the total rows in each table keeps unchanged.

In my test, *memory_usage* of DB increased from 1GB to 1.6 GB, which made sense due to the data update. However, the memory of the program increased from 2GB to 8.6GB. The latter is quite wired, and never happened in 0.5.1.

**Reply by Mytherin** [reply]
Are you using v0.6 or v0.6.1?

**Reply by foTok** [reply]
I used master-branch two days ago, not noticing v0.6.1 had been released. I will retest v0.6.1. Thanks.

**Reply by foTok** [reply]
> Are you using v0.6 or v0.6.1?

I tested v0.6.1 compiled by my self, all settings were default except that I commented *FSST* compression function. (I have to link jemalloc explicitly, or exception may occur at checkpoint).


DB used 1.9GB:
2022-12-08 10:01:54
database_size	block_size	total_blocks	used_blocks	free_blocks	wal_size	memory_usage	memory_limit	
VARCHAR	BIGINT	BIGINT	BIGINT	BIGINT	VARCHAR	VARCHAR	VARCHAR	
[ Rows: 1]
860.6MB	262144	3283	2872	411	234.2MB	1.9GB	53.9GB


Program used 8.8 GB：
2022-12-08 10:01:34
DuckDBVerification VSZmem: 8806848 VmRSS: 7023256 kB CPU: 317

**Reply by foTok** [reply]
> It is impossible to tell without knowing anything about what you are doing.


Hi, I found a simplest case:
1. Create a DB in file mode, create some table and load some original data.
2. Start a thread, which randomly insert and delete data in tables stored in file (also in a while loop): I always keep the insert number equals to the delete number, so the total rows in each table keeps unchanged.


Here is the results:
**At the beginning**:
con->Query("PRAGMA database_size")->Print()
```
 2022-12-09 12:00:00
database_size	block_size	total_blocks	used_blocks	free_blocks	wal_size	memory_usage	memory_limit	
VARCHAR	BIGINT	BIGINT	BIGINT	BIGINT	VARCHAR	VARCHAR	VARCHAR	
[ Rows: 1]
115.0MB	262144	439	360	79	191.0MB	1.1GB	13.3GB
```
shell
```
2022-12-09 12:00:00
DuckDBVerification VSZmem: 1799972 VmRSS: 1300888 kB CPU: 120
```

-------------------


**After 4 hours**:
con->Query("PRAGMA database_size")->Print()
```
2022-12-09 16:09:23
database_size	block_size	total_blocks	used_blocks	free_blocks	wal_size	memory_usage	memory_limit	
VARCHAR	BIGINT	BIGINT	BIGINT	BIGINT	VARCHAR	VARCHAR	VARCHAR	
[ Rows: 1]
342.6MB	262144	1307	1038	269	1.6MB	1.3GB	13.3GB
```

shell
```
2022-12-09 16:09:17
DuckDBVerification VSZmem: 5095224 VmRSS: 4508732 kB CPU: 127
```
--------------------------


**memory_usage of DB increased from 1.1 GB to 1.3 GB, however, the program used 5GB from 1.8GB.**

**Comment by l1t1** [comment]
delete from table didn't free the memory in duckdb. 
```
Report issues at https://termux.dev/issues
~ $ duckdb
v0.6.0 2213f9c
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D PRAGMA database_size;
┌───────────────┬────────────┬──────────────┬─────────────┬───┬──────────┬──────────────┬──────────────┐
│ database_size │ block_size │ total_blocks │ used_blocks │ … │ wal_size │ memory_usage │ memory_limit │
│    varchar    │   int64    │    int64     │    int64    │   │ varchar  │   varchar    │   varchar    │
├───────────────┼────────────┼──────────────┼─────────────┼───┼──────────┼──────────────┼──────────────┤
│ 0 bytes       │          0 │            0 │           0 │ … │ 0 bytes  │ 0 bytes      │ 6.4GB        │
├───────────────┴────────────┴──────────────┴─────────────┴───┴──────────┴──────────────┴──────────────┤
│ 1 rows                                                                           8 columns (7 shown) │
└──────────────────────────────────────────────────────────────────────────────────────────────────────┘
D
D create table t as select i from generate_series(1,10000000)t(i);
D
D
D PRAGMA database_size;
┌───────────────┬────────────┬──────────────┬─────────────┬───┬──────────┬──────────────┬──────────────┐
│ database_size │ block_size │ total_blocks │ used_blocks │ … │ wal_size │ memory_usage │ memory_limit │
│    varchar    │   int64    │    int64     │    int64    │   │ varchar  │   varchar    │   varchar    │
├───────────────┼────────────┼──────────────┼─────────────┼───┼──────────┼──────────────┼──────────────┤
│ 0 bytes       │          0 │            0 │           0 │ … │ 0 bytes  │ 106.9MB      │ 6.4GB        │
├───────────────┴────────────┴──────────────┴─────────────┴───┴──────────┴──────────────┴──────────────┤
│ 1 rows                                                                           8 columns (7 shown) │
└──────────────────────────────────────────────────────────────────────────────────────────────────────┘
D
D truncate table t;
100% ▕████████████████████████████████████████████████████████████▏
D
D PRAGMA database_size;
┌───────────────┬────────────┬──────────────┬─────────────┬───┬──────────┬──────────────┬──────────────┐
│ database_size │ block_size │ total_blocks │ used_blocks │ … │ wal_size │ memory_usage │ memory_limit │
│    varchar    │   int64    │    int64     │    int64    │   │ varchar  │   varchar    │   varchar    │
├───────────────┼────────────┼──────────────┼─────────────┼───┼──────────┼──────────────┼──────────────┤
│ 0 bytes       │          0 │            0 │           0 │ … │ 0 bytes  │ 106.9MB      │ 6.4GB        │
├───────────────┴────────────┴──────────────┴─────────────┴───┴──────────┴──────────────┴──────────────┤
│ 1 rows                                                                           8 columns (7 shown) │
└──────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

**Reply by Alex-Monahan** [reply]
Deletions do not free memory to the OS today. It should be reused by DuckDB though I believe.
We focus on analytical use cases not transactional ones. How is this an analytical use case?

**Reply by foTok** [reply]
> Deletions do not free memory to the OS today. It should be reused by DuckDB though I believe. We focus on analytical use cases not transactional ones. How is this an analytical use case?

1.  I employed file mode, which should not keep all data in memory. 
2.  The key point is, in my case, after some update, the "memory_usage" increased from 1.1 GB to 1.3 GB, however, the program's memory increased from 1.8 GB to 5GB.  Which means much memory was not managed by duckdb's BlockManager.
3. Although a deletion does not really remove the row, I think the row should be reused in the next insertition.

I also tested the case in v0.5.1, in which the program cost much less memory than 0.6.1

**Reply by foTok** [reply]
> delete from table didn't free the memory in duckdb.
> 
> ```
> Report issues at https://termux.dev/issues
> ~ $ duckdb
> v0.6.0 2213f9c
> Enter ".help" for usage hints.
> Connected to a transient in-memory database.
> Use ".open FILENAME" to reopen on a persistent database.
> D PRAGMA database_size;
> ┌───────────────┬────────────┬──────────────┬─────────────┬───┬──────────┬──────────────┬──────────────┐
> │ database_size │ block_size │ total_blocks │ used_blocks │ … │ wal_size │ memory_usage │ memory_limit │
> │    varchar    │   int64    │    int64     │    int64    │   │ varchar  │   varchar    │   varchar    │
> ├───────────────┼────────────┼──────────────┼─────────────┼───┼──────────┼──────────────┼──────────────┤
> │ 0 bytes       │          0 │            0 │           0 │ … │ 0 bytes  │ 0 bytes      │ 6.4GB        │
> ├───────────────┴────────────┴──────────────┴─────────────┴───┴──────────┴──────────────┴──────────────┤
> │ 1 rows                                                                           8 columns (7 shown) │
> └──────────────────────────────────────────────────────────────────────────────────────────────────────┘
> D
> D create table t as select i from generate_series(1,10000000)t(i);
> D
> D
> D PRAGMA database_size;
> ┌───────────────┬────────────┬──────────────┬─────────────┬───┬──────────┬──────────────┬──────────────┐
> │ database_size │ block_size │ total_blocks │ used_blocks │ … │ wal_size │ memory_usage │ memory_limit │
> │    varchar    │   int64    │    int64     │    int64    │   │ varchar  │   varchar    │   varchar    │
> ├───────────────┼────────────┼──────────────┼─────────────┼───┼──────────┼──────────────┼──────────────┤
> │ 0 bytes       │          0 │            0 │           0 │ … │ 0 bytes  │ 106.9MB      │ 6.4GB        │
> ├───────────────┴────────────┴──────────────┴─────────────┴───┴──────────┴──────────────┴──────────────┤
> │ 1 rows                                                                           8 columns (7 shown) │
> └──────────────────────────────────────────────────────────────────────────────────────────────────────┘
> D
> D truncate table t;
> 100% ▕████████████████████████████████████████████████████████████▏
> D
> D PRAGMA database_size;
> ┌───────────────┬────────────┬──────────────┬─────────────┬───┬──────────┬──────────────┬──────────────┐
> │ database_size │ block_size │ total_blocks │ used_blocks │ … │ wal_size │ memory_usage │ memory_limit │
> │    varchar    │   int64    │    int64     │    int64    │   │ varchar  │   varchar    │   varchar    │
> ├───────────────┼────────────┼──────────────┼─────────────┼───┼──────────┼──────────────┼──────────────┤
> │ 0 bytes       │          0 │            0 │           0 │ … │ 0 bytes  │ 106.9MB      │ 6.4GB        │
> ├───────────────┴────────────┴──────────────┴─────────────┴───┴──────────┴──────────────┴──────────────┤
> │ 1 rows                                                                           8 columns (7 shown) │
> └──────────────────────────────────────────────────────────────────────────────────────────────────────┘
> ```

Seems solved by #5698.
